### PR TITLE
Run unit tests on Chrome and Safari via Sauce Labs

### LINKED ISSUE #6039: Run unit tests on more browsers
Currently unit tests only run on latest Chrome and Chrome 45, we need to run them on more browsers, ideally all the browsers that run integration test. Unfortunately unit tests are so fined tuned to Chrome that making them run on all SauceLabs browsers will be a big task. Instead let's slowly fix them for different browser.

- [x] Remove `--oldchrome` (SL_Chrome_45)
- [x] Introduce `saucelabs-lite`, this will be a subset of SauceLabs browsers where unit test runs are stable and pass (Initially it can just include ~~SL_Chrome_45~~ SL_Chrome_latest)
- [x] Fix unit tests for Safari latest and add it to `saucelabs-lite`
- [ ] Fix unit tests for Chrome Android and add it to `saucelabs-lite`
- etc...

/cc @erwinmombay thoughts?

### PR DESCRIPTION
**Pull request by rsimha-amp**
With this PR, in addition to running all unit tests on Chrome (latest), we run a subset of the unit tests on Safari (latest) via Sauce Labs.

Fixes #6039

/to @choumx @erwinmombay @aghassemi

### CODE REVIEW

**Comment by choumx** [comment]
Is there value in running tests on Saucelabs + Chrome latest? I'm not sure we're actually getting additional signal here.

**Comment by rsimha-amp** [comment]
@choumx This is a prequel to the PRs that will enable the unit tests on non-chrome-latest browsers. Right now, there are a bunch of unit tests that fail when run on other browsers, so I'm going to have to do that clean up in a separate PR.

**Comment by choumx** [comment]
So we'd disable latest-Chrome once we add support for other browsers?

I guess it's not obvious to me that #6039 is a good idea in the first place. Will we need to start transpiling test code to get it to run on browsers other than latest Chrome?

**Comment by rsimha-amp** [comment]
@choumx You've raised a good point :)

We wouldn't disable unit tests on latest chrome on sauce labs, since this PR will no longer run unit tests on the version of chrome that's native to Travis. The idea is to (soon) run unit tests on latest chrome, latest safari, and latest chrome android.

I don't believe we'll need to transpile the tests on a couple of the latest browsers. The intention isn't to run unit tests on the entire sauce labs browsers.

@aghassemi can comment on the rationale underlying his original issue, after which we can determine if this PR is still a good idea.

**Comment by aghassemi** [comment]
@choumx @choumx The rational behind https://github.com/ampproject/amphtml/issues/6039 is that we had a few production bugs due to code using web platforms API's not available in Safari. Unit tests would have caught those bugs but they do not run on Safari. So getting unit tests run in latest Safari same way we do in latest Chrome is the important part of this. Other browsers, older versions is a nice to have.

**Comment by choumx** [comment]
@aghassemi Integration testing + presubmit bans for APIs with cross-browser issues seems better IMO. We already have infra for both and we won't need to polyfill/transpile test code.

**Comment by rsimha-amp** [comment]
@choumx @aghassemi As of now, there are ~20 failures when unit tests are run on Safari: https://travis-ci.org/ampproject/amphtml/jobs/293802513

I expect that with a few fixes, we should be able to run unit tests on Safari for all PRs (or at least for push builds). Since tests on different browsers are run in parallel, I expect that the overall increase in build time should be small relative to the increase in the number of tests we run.

**Comment by rsimha-amp** [comment]
@aghassemi @choumx Is there a list of APIs with cross-browser issues? Would it make sense to ban those APIs via presubmit checks instead of running unit tests on Safari?

**Comment by aghassemi** [comment]
@rsimha-amp @choumx The idea behind running unit tests in Safari was to "catch" these cross-browser issues that devs initially didn't realize is an issue and hence is not banned in presubmit. After we know they are an issue, then yes banning them is the right action. I see having unit tests run in Safari as just anther safety check to increases chances of catching issues we are not aware of.

**Comment by rsimha-amp** [comment]
@aghassemi @choumx I've gone ahead and added a new commit to this PR that fixes / skips the individual tests that currently fail on Safari, and assigned TODOs to the owners of each of those tests. I've also enabled unit test runs on Chrome and Safari. Hopefully, the next Travis PR check build should be green.

When I looked at the overall timing of the build, there wasn't a significant increase in duration, due to the fact that tests are run in parallel on the two browsers. With this, I'm hoping both your concerns should be addressed.

PTAL.

**Comment by rsimha-amp** [comment]
I saw some disconnect errors between the Karma runner and the Safari browser instance on Sauce Labs. This is a known issue: https://support.saucelabs.com/hc/en-us/articles/115010079868-Issues-with-Safari-and-Karma-Test-Runner

I've added a commit that should hopefully fix this.

**Comment by aghassemi** [comment]
@rsimha-amp unit tests are still failing, will review after green?

**Comment by rsimha-amp** [comment]
@aghassemi Running a subset of the tests on Safari works. See https://travis-ci.org/ampproject/amphtml/jobs/302046377

This is ready for a full review. You can use https://github.com/ampproject/amphtml/pull/11827/files?w=1 to filter out whitespace-only changes while reviewing.

**Comment by rsimha-amp** [comment]
@aghassemi Ping

**Comment by rsimha-amp** [comment]
@erwinmombay could you take a look? I have a couple other PRs that are waiting for this to be merged.

File: test/functional/test-custom-element.js [hunk 1]
```diff
@@ -26,1295 +26,1299 @@ import * as lolex from 'lolex';
...
-      testElementUnlayoutCallback();
-      return true;
-    }
-    pauseCallback() {
-      testElementPauseCallback();
-    }
-    resumeCallback() {
-      testElementResumeCallback();
+  // TODO(dvoytenko): Make this test work on Safari. See #11827.
+  describe.configure().skipSafari().run('CustomElement', () => {
```
Comment (aghassemi) [inline test/functional/test-custom-element.js]: GitHub UI doesn't make it clear what actually changed here. Is there anything changes other than `skipSafari`?

File: test/functional/test-ampdoc.js [hunk 2]
```diff
@@ -166,7 +166,8 @@ describe('AmpDocService', () => {
       }).to.throw(/The shadow root already contains ampdoc/);
     });
 
-    it('should navigate via host', () => {
+    // TODO(dvoytenko): Make this test work on Safari. See #11827.
```
Comment (aghassemi) [inline test/functional/test-ampdoc.js]: nit: `// TODO(dvoytenko, #11827)` (Dima has a tool that extracts these if the TODO(username, #issue) format is followed.

File: test/functional/test-custom-element.js (hunk 1 above)
Comment (rsimha-amp) [reply test/functional/test-custom-element.js]: No, this just adds the `skipSafari`.

Did you click the link I sent you? https://github.com/ampproject/amphtml/pull/11827/files?w=1 Adding `?w=1` will filter out whitespace changes.

File: test/functional/test-ampdoc.js (hunk 2 above)
Comment (rsimha-amp) [reply test/functional/test-ampdoc.js]: Fixed.

File: test/functional/test-custom-element.js (hunk 1 above)
Comment (aghassemi) [reply test/functional/test-custom-element.js]: Whoa! had no idea about `?w=1` stealing it for the next time I am assigned tip of the week.

**Pull request merged**